### PR TITLE
Feature/numeric stepper

### PR DESCRIPTION
**Pull request by dramoturg**

### CODE REVIEW

File: src/components/Input/Input.demo.styl [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+	&_isFocused,
+	&:focus {
+		color: #000;
+		background: #fff;
+		border-color: #a9a9a9;
+		box-shadow: none;
+	}
+
+	&_isReadonly,
```
Comment (raveclassic) [inline src/components/Input/Input.demo.styl]: something one should be left, "&:read-only" is preferable

File: src/components/NumericStepper/NumericStepper.jsx [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+	static propTypes = {
+		...Input.propTypes,
+		step: React.PropTypes.number,
+		min: React.PropTypes.number,
+		max: React.PropTypes.number,
+		formatter: React.PropTypes.func,
+		ButtonComponent: React.PropTypes.func,
+		InputComponent: React.PropTypes.func,
+
+		repeatIntervar: React.PropTypes.number,
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: typo

File: src/components/NumericStepper/NumericStepper.jsx [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+		if (!isChanged && isCorrected) {
+			this.setState({
+				isCorrected
+			});
+		}
+	}
+
+	getCorrectedValue(value) {
+		const {min, max} = this.props;
+		let newValue = parseFloat(value);
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: don't use parseFloat as it produces false-positive results for values like '123abc'

File: src/components/NumericStepper/NumericStepper.jsx [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+	}
+
+	componentWillUpdate(newProps, newState) {
+		if (newState.isCorrected) {
+			const timout = setTimeout(() => {
+				this.setState({
+					isCorrected: !newState.isCorrected
+				});
+				clearTimeout(timout);
+			}, 750);
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: why do you need timeout here?

File: src/components/NumericStepper/NumericStepper.jsx [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+	onButtonDownClick = () => {
+		this.decrease();
+	}
+
+	onButtonUpClick = () => {
+		this.increase();
+	}
+}
+
+class Holdable extends React.Component {
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: Let's move it to separate file

File: src/components/NumericStepper/NumericStepper.jsx [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+			onFocus,
+			pattern,
+			onKeyDown,
+			theme: inputTheme
+		};
+
+		if (!isDisabled && isFocused) {
+			inputProps.onWheel = onWheel;
+		}
+		const buttonTheme = Object.keys(BUTTON_TYPE).reduce((acc, current) => {
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: button themes can be constructed at module-level

File: src/components/NumericStepper/NumericStepper.jsx [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+				}
+			};
+		}, {});
+
+		return (
+			<div className={theme.container}>
+				<Input {...inputProps} />
+				<div className={theme.container__buttons}>
+					<Holdable onHold={this.onButtonDownClick}
+							  isDisabled={isDisabled || value <= min}>
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: try to update webstorm and check 'smart tabs' should be enabled

File: src/components/Input/Input.demo.styl (hunk 1 above)
Comment (dramoturg) [inline src/components/Input/Input.demo.styl]: &:read-only selector don`t support ie, firefox.  
image

File: src/components/NumericStepper/NumericStepper.jsx (hunk 6 above)
Comment (dramoturg) [inline src/components/NumericStepper/NumericStepper.jsx]: button themes uses `theme` props.

File: package.json [hunk 8]
```diff
@@ -7,6 +7,7 @@
     "bower": "bower",
     "eslint": "eslint --config ./node_modules/dx-lint/.react.eslintrc --ext .js,.jsx .",
     "start": "node ./scripts/run.js ./scripts/start.js",
+    "postinstall": "bower prune && bower install",
```
Comment (raveclassic) [inline package.json]: merge develop

File: src/components/Input/Input.demo.styl (hunk 1 above)
Comment (raveclassic) [inline src/components/Input/Input.demo.styl]: let's take `[readonly]`, `:focus`, `:disabled`

File: src/components/Input/Input.jsx [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+	static propTypes = PROP_TYPES;
+
+	constructor(...args) {
+		super(...args);
+		const {value} = this.props;
+
+		this.state = {
+			value,
+			isFocused: false
```
Comment (raveclassic) [inline src/components/Input/Input.jsx]: why do you need isFocused state? input styles can be handled via `:focus`

File: src/components/NumericStepper/NumericStepper.jsx (hunk 5 above)
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: Try to `mousedown`, then drag without release and then `mouseup`. Incrementation/Decrementation is cycled.

File: src/components/NumericStepper/NumericStepper.jsx [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+	decrease = () => {
+		this.step(-1);
+	}
+
+	componentWillReceiveProps(newProps) {
+		this.setValue(newProps.value);
+	}
+
+	componentWillUpdate(newProps, newState) {
+		if (newState.isCorrected) {
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: if you get called multiple times here these timeouts can potentially end with a race condition

File: src/components/NumericStepper/NumericStepper.jsx [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+	UP: 38,
+	DOWN: 40
+};
+
+@PURE
+@themr(NUMERIC_STEPPER)
+export default class NumericStepper extends Component {
+
+	static propTypes = {
+		...Input.propTypes,
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: why do you mix in these propTypes? numeric stepper is not a kind of input, it's a separate component

File: src/components/NumericStepper/NumericStepper.demo.styl [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+		}
+		50% {
+			background-color: #7f0d0d;
+		}
+		100% {
+			background-color: #202020;
+		}
+	}
+
+	@keyframes corrected {
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.demo.styl]: formatting

File: src/components/NumericStepper/NumericStepper.jsx [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+
+		repeatIntervar: React.PropTypes.number,
+		repeatDelay: React.PropTypes.number,
+
+		upIconName: React.PropTypes.string,
+		downIconName: React.PropTypes.string,
+
+		theme: React.PropTypes.shape({
+			container: React.PropTypes.string,
+			container__input: React.PropTypes.string,
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: what's the point in prefixing?

File: src/components/NumericStepper/NumericStepper.jsx [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+};
+
+const KEYCODE = {
+	UP: 38,
+	DOWN: 40
+};
+
+@PURE
+@themr(NUMERIC_STEPPER)
+export default class NumericStepper extends Component {
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: There's no way to edit value in input when trying to remove symbols with backspace

Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: cannot edit value by typing at all, onChange is not triggered

**Comment by raveclassic** [comment]
Controlled behavior (when `value` is set from parent component) should be the same as for Selectbox. There's no need to set state and rerender if value is controlled from the outside.

**Comment by raveclassic** [comment]
Validation is run on initial render when value is not in passed min/max bounds. Validation should be run on touched (dirty) components and on focusout.

File: src/components/NumericStepper/NumericStepper.jsx [hunk 15]
```diff
@@ -0,0 +1,330 @@
...
+			<div className={theme.container}>
+				<Input {...inputProps} />
+				<div className={theme.container__buttons}>
+					<Holdable onHold={this.onButtonDownClick}
+							  isDisabled={isDisabled || value <= min}>
+						<Button onClick={this.onButtonDownClick}
+								name={downIconName}
+								theme={buttonTheme.UP} />
+					</Holdable>
+					<Holdable onHold={this.onButtonUpClick} isDisabled={isDisabled || value >= max}>
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: min/max bounds include the values actually, so check should `value > max`

File: src/components/NumericStepper/NumericStepper.jsx [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+			isCorrected: false,
+			isFocused: false
+		};
+	}
+
+	static defaultProps = {
+		step: 1,
+		max: Infinity,
+		min: 0,
+		value: 1,
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: whe the value is 1 by default? it should be required in propTypes and paired with defaultValue for uncontrolled behavior

**Comment by raveclassic** [comment]
Set max to 20 and step to 40 and increment. The behaviour is confusing as NumericStepper should fail validation and not set the max value.

File: src/components/NumericStepper/NumericStepper.jsx [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+		...Input.propTypes,
+		step: React.PropTypes.number,
+		min: React.PropTypes.number,
+		max: React.PropTypes.number,
+		formatter: React.PropTypes.func,
+		ButtonComponent: React.PropTypes.func,
+		InputComponent: React.PropTypes.func,
+
+		repeatIntervar: React.PropTypes.number,
+		repeatDelay: React.PropTypes.number,
```
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: what are these values for?

File: src/components/NumericStepper/NumericStepper.jsx (hunk 5 above)
Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: @mankdev Do we really need such complex behavior for now? Mousedowns/mouseups are incomplete as tap events are not supported.

File: src/components/NumericStepper/NumericStepper.jsx (hunk 17 above)
Comment (dramoturg) [inline src/components/NumericStepper/NumericStepper.jsx]: These props are passed to the component Holdable.

Comment (raveclassic) [inline src/components/NumericStepper/NumericStepper.jsx]: Then they should probably be not. Instead it's better to have an ability to pass down custom Holdable

File: src/components/Holdable/Holdable.jsx [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+				onHold && onHold();
+			}, interval);
+
+			this.forceUpdate();
+		}, delay);
+	}
+
+	stop() {
+		if (this._timeoutId) {
+			clearTimeout(this._timeoutId);
```
Comment (raveclassic) [inline src/components/Holdable/Holdable.jsx]: It's ok for now but actually both `clearTimeout` and `clearInverval` accept nulls so checking is redundant

**Comment by raveclassic** [comment]
still can't properly edit the value in input: 1. focus, 2. move caret to the end, 3. press backspace, 4. press backspace again - fail. 5. move caret one symbol on the left. 6. press delete - fail.

**Pull request merged**